File: dsa-python/solutions/trees/easy/diameter_of_binary_tree.py
# 543. Diameter of Binary Tree
# https://leetcode.com/problems/diameter-of-binary-tree/
#
# Given the root of a binary tree, return the length of the diameter of the tree.
# The diameter is the longest path between any two nodes (counted in edges).
# This path may or may not pass through the root.
#
# APPROACH: Postorder recursion. At each node, compute left and right heights.
# The path THROUGH this node = left_height + right_height.
# Track the max globally. Return height (1 + max(left, right)) to parent.
#
# TIME: O(n) — visit every node once
# SPACE: O(h) — recursion stack (h = height)


class TreeNode:
    def __init__(self, val: int = 0, left: "TreeNode | None" = None, right: "TreeNode | None" = None):
        self.val = val
        self.left = left
        self.right = right


class DiameterOfBinaryTree:
    def diameter_of_binary_tree(self, root: TreeNode | None) -> int:
        max_diameter = 0

        def height(node):
            nonlocal max_diameter
            if not node:
                return 0

            left = height(node.left)    # height of left subtree
            right = height(node.right)  # height of right subtree

            # Path through this node = left + right (update global max)
            max_diameter = max(max_diameter, left + right)

            # Return height of this node to parent
            return 1 + max(left, right)

        height(root)
        return max_diameter

File: dsa-csharp/solutions/Trees/Easy/DiameterOfBinaryTree.cs
// 543. Diameter of Binary Tree
// https://leetcode.com/problems/diameter-of-binary-tree/
//
// Given the root of a binary tree, return the length of the diameter.
// Diameter = longest path between any two nodes (counted in edges).
//
// APPROACH: Postorder recursion. At each node, compute left and right heights.
// Path THROUGH this node = left_height + right_height. Track max globally.
// Return height (1 + max(left, right)) to parent for its calculation.
//
// TIME: O(n) — visit every node once
// SPACE: O(h) — recursion stack (h = height)

public class DiameterOfBinaryTree
{
    public int DiameterOfBinaryTreeMethod(TreeNode? root)
    {
        int maxDiameter = 0;

        Height(root, ref maxDiameter);

        return maxDiameter;
    }

    private int Height(TreeNode? root, ref int maxDiameter)
    {
        if(root == null) return 0;

        int left = Height(root.left, ref maxDiameter);
        int right = Height(root.right, ref maxDiameter);

        maxDiameter = Math.Max(maxDiameter, left + right);

        return 1 + Math.Max(left, right);
    }
}
